File: go/book.go
package main

import "github.com/google/btree"

// ---------------------------------------------------------------- the book
// Price-time priority; lazy cancellation (cancel removes from the id map,
// stale ids are skipped at the head of a level queue).

type resting struct {
	price int64
	qty   int64
}

// FIFO of order ids as a slice with a head index.
type level struct {
	ids  []uint64
	head int
}

func (l *level) empty() bool    { return l.head == len(l.ids) }
func (l *level) front() uint64  { return l.ids[l.head] }
func (l *level) popFront()      { l.head++ }
func (l *level) push(id uint64) { l.ids = append(l.ids, id) }

type book struct {
	bidPrices *btree.BTreeG[int64]
	askPrices *btree.BTreeG[int64]
	bidLevels map[int64]*level
	askLevels map[int64]*level
	orders    map[uint64]resting
	fills     uint64
	volume    uint64
}

func newBook() *book {
	less := func(a, b int64) bool { return a < b }
	return &book{
		bidPrices: btree.NewG(16, less),
		askPrices: btree.NewG(16, less),
		bidLevels: make(map[int64]*level),
		askLevels: make(map[int64]*level),
		orders:    make(map[uint64]resting),
	}
}

func (b *book) limit(side byte, price, qty int64, id uint64) {
	for qty > 0 {
		// best opposite level that crosses, else rest the residual
		var bestPrice int64
		var lvl *level
		if side == sideBid {
			p, ok := b.askPrices.Min()
			if !ok || p > price {
				break
			}
			bestPrice, lvl = p, b.askLevels[p]
		} else {
			p, ok := b.bidPrices.Max()
			if !ok || p < price {
				break
			}
			bestPrice, lvl = p, b.bidLevels[p]
		}
		for qty > 0 {
			if lvl.empty() {
				break
			}
			head := lvl.front()
			rec, ok := b.orders[head]
			if !ok { // lazily-cancelled order
				lvl.popFront()
				continue
			}
			m := min(qty, rec.qty)
			rec.qty -= m
			qty -= m
			b.fills++
			b.volume += uint64(m)
			if rec.qty == 0 {
				delete(b.orders, head)
				lvl.popFront()
			} else {
				b.orders[head] = rec
			}
		}
		if lvl.empty() {
			if side == sideBid {
				b.askPrices.Delete(bestPrice)
				delete(b.askLevels, bestPrice)
			} else {
				b.bidPrices.Delete(bestPrice)
				delete(b.bidLevels, bestPrice)
			}
		}
	}
	if qty > 0 {
		prices, levels := b.bidPrices, b.bidLevels
		if side != sideBid {
			prices, levels = b.askPrices, b.askLevels
		}
		lvl, ok := levels[price]
		if !ok {
			lvl = &level{}
			levels[price] = lvl
			prices.ReplaceOrInsert(price)
		}
		lvl.push(id)
		b.orders[id] = resting{price: price, qty: qty}
	}
}

func (b *book) cancel(id uint64) {
	delete(b.orders, id)
}

func (b *book) stats() (uint64, uint64, int, uint64) {
	return b.fills, b.volume, len(b.orders), b.checksum()
}

// checksum over live orders in (price ascending, FIFO) order per side —
// must be identical to the Java and Rust runs on the same workload.
func (b *book) checksum() uint64 {
	var acc uint64
	fold := func(prices *btree.BTreeG[int64], levels map[int64]*level) {
		prices.Ascend(func(p int64) bool {
			lvl := levels[p]
			for _, id := range lvl.ids[lvl.head:] {
				if r, ok := b.orders[id]; ok {
					s := splitMix64{acc ^ id ^ (uint64(r.price) << 20) ^ uint64(r.qty)}
					acc = s.next()
				}
			}
			return true
		})
	}
	fold(b.bidPrices, b.bidLevels)
	fold(b.askPrices, b.askLevels)
	return acc
}

File: rust/src/book.rs
//! The idiomatic price-time priority book.

use std::collections::{BTreeMap, HashMap, VecDeque};

use crate::workload::{SplitMix64, SIDE_BID};

// ---------------------------------------------------------------- the book
// Price-time priority. Lazy cancellation: cancel removes from the id map;
// stale ids are skipped when they reach the head of a level queue.

#[derive(Clone, Copy)]
pub struct Resting {
    pub price: i64,
    pub qty: i64,
}

#[derive(Default)]
pub struct Book {
    pub bids: BTreeMap<i64, VecDeque<u64>>,
    pub asks: BTreeMap<i64, VecDeque<u64>>,
    pub orders: HashMap<u64, Resting>,
    pub fills: u64,
    pub volume: u64,
}

impl Book {
    pub fn limit(&mut self, side: u8, price: i64, mut qty: i64, id: u64) {
        while qty > 0 {
            // best opposite level that crosses, else rest the residual
            let (best_price, level) = if side == SIDE_BID {
                match self.asks.iter_mut().next() {
                    Some((p, l)) if *p <= price => (*p, l),
                    _ => break,
                }
            } else {
                match self.bids.iter_mut().next_back() {
                    Some((p, l)) if *p >= price => (*p, l),
                    _ => break,
                }
            };
            while qty > 0 {
                let Some(&head) = level.front() else { break };
                let Some(rec) = self.orders.get_mut(&head) else {
                    level.pop_front(); // lazily-cancelled order
                    continue;
                };
                let m = qty.min(rec.qty);
                rec.qty -= m;
                qty -= m;
                self.fills += 1;
                self.volume += m as u64;
                if rec.qty == 0 {
                    self.orders.remove(&head);
                    level.pop_front();
                }
            }
            if level.is_empty() {
                if side == SIDE_BID {
                    self.asks.remove(&best_price);
                } else {
                    self.bids.remove(&best_price);
                }
            }
        }
        if qty > 0 {
            let book_side = if side == SIDE_BID { &mut self.bids } else { &mut self.asks };
            book_side.entry(price).or_default().push_back(id);
            self.orders.insert(id, Resting { price, qty });
        }
    }

    pub fn cancel(&mut self, id: u64) {
        self.orders.remove(&id);
    }

    /// Deterministic checksum over live orders in (price, FIFO) order — must be
    /// identical between the Rust and Java runs on the same workload.
    pub fn checksum(&self) -> u64 {
        let mut acc = 0u64;
        let fold = |acc: &mut u64, id: u64, r: &Resting| {
            let mut h = SplitMix64(*acc ^ id ^ ((r.price as u64) << 20) ^ (r.qty as u64));
            *acc = h.next();
        };
        for level in self.bids.values() {
            for &id in level {
                if let Some(r) = self.orders.get(&id) {
                    fold(&mut acc, id, r);
                }
            }
        }
        for level in self.asks.values() {
            for &id in level {
                if let Some(r) = self.orders.get(&id) {
                    fold(&mut acc, id, r);
                }
            }
        }
        acc
    }
}

